api/deployment/DeploymentConfigRestHandler.go

```diff
@@ -113,12 +117,73 @@ func (handler *DeploymentConfigRestHandlerImpl) CreateChartFromFile(w http.Respo
 		},
 	}
 
-	err = handler.chartRefRepository.Save(chartRefs)
+	chartsJson, _ := json.Marshal(chartRefs)
+	err = ioutil.WriteFile(filepath.Join(chartInfo.TemporaryFolder, "output.json"), chartsJson, 0644)
+	if err != nil {
+		common.WriteJsonResp(w, fmt.Errorf(err.Error()), nil, http.StatusInternalServerError)
+		return
+	}
+
+	pathList := strings.Split(chartInfo.TemporaryFolder, "/")
+	chartData := &DeploymentChartInfo{
+		ChartName:   chartInfo.ChartName,
+		Description: chartInfo.Description,
+		FileId:      pathList[len(pathList)-1],
+	}
+
+	common.WriteJsonResp(w, err, chartData, http.StatusOK)
+	return
+}
+
+func (handler *DeploymentConfigRestHandlerImpl) SaveChart(w http.ResponseWriter, r *http.Request) {
+	decoder := json.NewDecoder(r.Body)
+	userId, err := handler.userAuthService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		common.WriteJsonResp(w, err, nil, http.StatusUnauthorized)
+		return
+	}
+
+	token := r.Header.Get("token")
+	if ok := handler.enforcer.Enforce(token, casbin.ResourceGlobal, casbin.ActionDelete, "*"); !ok {
+		common.WriteJsonResp(w, errors.New("unauthorized"), nil, http.StatusForbidden)
+		return
+	}
+
+	var request DeploymentChartInfo
+	err = decoder.Decode(&request)
 	if err != nil {
-		handler.Logger.Errorw("error in saving ConfigMap, CallbackConfigMap", "err", err)
-		common.WriteJsonResp(w, err, "Chart couldn't be saved", http.StatusInternalServerError)
+		handler.Logger.Errorw("decode err", "err", err)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
 		return
 	}
-	common.WriteJsonResp(w, err, "Chart Saved Successfully", http.StatusOK)
+
+	location := filepath.Join(string(handler.refChartDir), request.FileId)
+	if request.Action == "Save" {
+		file, _ := ioutil.ReadFile(filepath.Join(location, "output.json"))
+		chartRefs := &chartRepoRepository.ChartRef{}
+		err = json.Unmarshal(file, &chartRefs)
+		if err != nil {
+			handler.Logger.Errorw("unmarshall err", "err", err)
+			common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+			return
+		}
+
+		err = handler.chartRefRepository.Save(chartRefs)
```

Comment:
we should validate again as someone can call this api even if result of previous api call was error

Reply:
FieldId is sent only when upload doesnt have error, hence not required to check again